File: LAB2/C#/Lab2Pop/Lab2Pop/ArrClass.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

namespace Lab2Pop
{
    public class ArrClass
    {
        private int _countOfThreads;
        private int[] _arr;

        private object lockerForCount = new object();
        private object lockerForMin = new object();
        private int _threadCount;
        int ArrMinIndex = 0;

        public ArrClass(int countOfThreads, int[] arr)
        {
            _countOfThreads = countOfThreads;
            _arr = arr;
        }

        public int FindMin(int startIndex, int endIndex)
        {
            int min = int.MaxValue;
            int index = 0;

            for (int i = startIndex; i < endIndex; i++)
            {
                if (min > _arr[i])
                {
                    min = _arr[i];
                    index = i;
                }
            }

            return index;
        }

        public int ParallelFindMin()
        {
            
            Thread[] threads = new Thread[_countOfThreads];
            int len = _arr.Length / _countOfThreads;

            for (int i = 0; i < _countOfThreads; i++)
            {
                int startIndex = len * i;
                int endIndex = (i == _countOfThreads - 1) ? _arr.Length : len * (i + 1);
                threads[i] = new Thread(StartThread);
                threads[i].Start(new Border(startIndex, endIndex));
            }

            lock (lockerForCount)
            {
                while (_threadCount < _countOfThreads)
                {
                    Monitor.Wait(lockerForCount);
                }
            }

            return ArrMinIndex;
        }

        private void StartThread(object param)
        {
            if (param is Border)
            {
                int minIndex = FindMin((param as Border).StartIndex,(param as Border).EndIndex);

                lock (lockerForMin)
                {
                    SetMinIndex(minIndex);
                }
             
                IncreaseThreadCount();
            }
        }

        private void SetMinIndex(int minIndex)
        {
            if (_arr[ArrMinIndex] > _arr[minIndex])
            {
                ArrMinIndex = minIndex;
            }
        }

        private void IncreaseThreadCount()
        {
            lock (lockerForCount)
            {
                _threadCount++;
                Monitor.Pulse(lockerForCount);
            }
        }
    }
}

File: LAB2/Java/Lab2Pop/src/ArrClass.java
import java.util.Random;

public class ArrClass {
    private final int _threadNum;
    public final int[] _arr;
    private int _threadCount = 0;
    private int _minIndex = 0;

    public ArrClass(int threadNum, int[] arr) {
        _threadNum = threadNum;
        _arr = arr;
    }

    public int findMinIndex(int startIndex, int finishIndex){
        int min = Integer.MAX_VALUE;
        int index = 0;
        for(int i = startIndex; i < finishIndex; i++){
            if (min > _arr[i])
            {
                min = _arr[i];
                index = i;
            }
        }
        return index;
    }

    synchronized private int getMinIndex() {
        while (getThreadCount()< _threadNum){
            try {
                wait();
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
        return _minIndex;
    }

    synchronized public void SetMinIndex(int minIndex){
        if (_arr[_minIndex] > _arr[minIndex]){
            _minIndex = minIndex;
        }
    }


    synchronized public void incThreadCount(){
        _threadCount++;
        notify();
    }

    private int getThreadCount() {
        return _threadCount;
    }

    public int parallelFindMin(){
        ThreadMin[] threadMins = new ThreadMin[_threadNum];
        int len = _arr.length / _threadNum;

        for (int i = 0; i < _threadNum; i++)
        {
            int startIndex = len * i;
            int endIndex = (i == _threadNum - 1) ? _arr.length : len * (i + 1);
            threadMins[i] = new ThreadMin(startIndex,endIndex, this);
            threadMins[i].start();
        }
        return getMinIndex();
    }
}
